collectors/htlcs_collector.go

```diff
@@ -0,0 +1,209 @@
+// waits for them to exit.
+func (h *htlcMonitor) stop() {
+	htlcLogger.Info("Stopping Htlc Monitor")
+
+	close(h.quit)
+	h.wg.Wait()
+}
+
+// collectors returns all of the collectors that the htlc monitor uses.
+func (h *htlcMonitor) collectors() []prometheus.Collector {
+	return []prometheus.Collector{
+		h.successCounter, h.failureCounter,
+	}
+}
+
+// consumeHtlcEvents subscribes to a stream of htlc events from lnd and records
+// running totals for our htlcs.
+func (h *htlcMonitor) consumeHtlcEvents() error {
+	// Create a context to subscribe to events and cancel it on exit so that
+	// lnd can cancel the stream.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	htlcEvents, htlcErrChan, err := h.router.SubscribeHtlcEvents(ctx)
+	if err != nil {
+		return err
+	}
+
+	h.wg.Add(1)
+	go func() {
+		defer func() {
+			cancel()
+			h.wg.Done()
+		}()
+
+		for {
+			select {
+			case event, ok := <-htlcEvents:
+				if !ok {
+					h.errChan <- errors.New("htlc event " +
+						"stream termianted")
+					return
+				}
+
+				err := h.processHtlcEvent(event)
+				if err != nil {
+					h.errChan <- err
+					return
+				}
+
+			case err, ok := <-htlcErrChan:
+				h.errChan <- fmt.Errorf("htlc stream "+
+					"exited: %v, closed: %v", err, ok)
+				return
+
+			case <-h.quit:
+				h.errChan <- errors.New("htlc collector " +
+					"shutting down")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// processHtlcEvent processes all the htlc events we consume from our stream.
+func (h *htlcMonitor) processHtlcEvent(event interface{}) error {
+	switch e := event.(type) {
+	// If we have received a forwarding event, we add it to our map if it
+	// is not already present. We are ok with duplicate events, because
+	// htlcs are sometimes replayed by the switch, but we want to keep our
+	// earliest timestamp for stats.
+	case *htlcswitch.ForwardingEvent:
+		if _, ok := h.activeForwards[e.HtlcKey]; ok {
+			htlcLogger.Infof("htlc: %v replayed", e.HtlcKey)
+			return nil
+		}
+
+		// If this forward is from one of our own sends, we just
+		// ignore it.
+		if e.HtlcEventType != htlcswitch.HtlcEventTypeForward {
+			return nil
+		}
+
+		// Add to our set of known forwards.
+		h.activeForwards[e.HtlcKey] = struct{}{}
+
+	case *htlcswitch.SettleEvent:
+		h.recordResolution(e.HtlcKey, e.HtlcEventType, true)
+
+	case *htlcswitch.ForwardingFailEvent:
+		h.recordResolution(e.HtlcKey, e.HtlcEventType, false)
+
+	case *htlcswitch.LinkFailEvent:
+		h.recordResolution(e.HtlcKey, e.HtlcEventType, false)
+
+	default:
+		return fmt.Errorf("unknown htlc event type: %T", event)
+	}
+
+	return nil
+}
+
+// recordResolution records the outcome of a htlc resolution (settle/fail) in
+// our metrics.
+func (h *htlcMonitor) recordResolution(key htlcswitch.HtlcKey,
+	eventType htlcswitch.HtlcEventType, success bool) {
+
+	// If the event is not related to a forward, we have no need to add it.
+	if eventType != htlcswitch.HtlcEventTypeForward {
```

Comment:
Shouldn't this check the opposite? That we only care about non forward events in this case.

Reply:
This type can be `forward`/`send`/`receive` to help users tell whether the htlc is theirs or not, so here we're just filtering out our own stuff which I imagine we don't care about for this dashboard?

Added in original lnd PR because we don't have the info to tell the difference between a failed receive and a forward which is failed on our incoming link (both have no outgoing chan ID).